### PR TITLE
ESQL: Impersonate an old client in upgrade tests

### PR DESCRIPTION
**Pull request by nik9000**
In the ESQL quests for upgrading from older versions of Elasticsearch this will pretend to be and old version of the elasticsearch client and not send a `version` in the body. Old clients will instead default to the first version of ESQL. This should allow us to run these tests even when `version` is required.

Tests against newer versions of Elasticsearch will then specify a `version`.

### CODE REVIEW

File: x-pack/plugin/core/template-resources/src/main/resources/monitoring-es-mb.json [hunk 1]
```diff
@@ -2366,7 +2366,7 @@
                 "shards_stats": {
                   "properties": {
                     "total_count": {
-                      "type": long
+                      "type": "long"
```
Comment (nik9000) [inline x-pack/plugin/core/template-resources/src/main/resources/monitoring-es-mb.json:2369]: This is already in main actually. Just ignore it. It'll vanish.

File: x-pack/plugin/esql/qa/server/mixed-cluster/src/yamlRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/EsqlClientYamlIT.java [hunk 2]
```diff
@@ -32,12 +44,59 @@ public EsqlClientYamlIT(final ClientYamlTestCandidate testCandidate) {
...
+                );
+                result.add(new Object[] { new ClientYamlTestCandidate(candidate.getRestTestSuite(), modified) });
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("error modifying " + candidate + ": " + e.getMessage(), e);
+            }
+        }
+        return result;
+    }
+
+    private static ExecutableSection modifyExecutableSection(ExecutableSection e) {
```
Comment (nik9000) [inline x-pack/plugin/esql/qa/server/mixed-cluster/src/yamlRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/EsqlClientYamlIT.java:69]: Remove `version` from all esql requests if we're running against before 8.14

File: x-pack/plugin/esql/qa/server/mixed-cluster/src/yamlRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/EsqlClientYamlIT.java [hunk 3]
```diff
@@ -32,12 +44,59 @@ public EsqlClientYamlIT(final ClientYamlTestCandidate testCandidate) {
...
     }
 
     @Before
     @After
     public void assertRequestBreakerEmpty() throws Exception {
         EsqlSpecTestCase.assertRequestBreakerEmpty();
     }
+
+    @Override
+    protected ClientYamlTestClient initClientYamlTestClient(
```
Comment (nik9000) [inline x-pack/plugin/esql/qa/server/mixed-cluster/src/yamlRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/EsqlClientYamlIT.java:92]: Send the old client signature if we're before 8.14.

**Comment by nik9000** [comment]
run elasticsearch-ci/part-2

File: x-pack/plugin/esql/qa/server/mixed-cluster/src/yamlRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/EsqlClientYamlIT.java [hunk 4]
```diff
@@ -32,12 +44,59 @@ public EsqlClientYamlIT(final ClientYamlTestCandidate testCandidate) {
...
+        }
+        DoSection doSection = (DoSection) e;
+        String api = doSection.getApiCallSection().getApi();
+        if (false == api.equals("esql.query")) {
+            return e;
+        }
+        for (Map<String, Object> body : doSection.getApiCallSection().getBodies()) {
+            body.remove("version");
+            // TODO if the version isn't the earliest we should skip the test - ES isn't going to support it.
+        }
```
Comment (alex-spies) [inline x-pack/plugin/esql/qa/server/mixed-cluster/src/yamlRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/EsqlClientYamlIT.java:81]: This is cool, thanks!

**Review by alex-spies — APPROVED** [review]
This is great, thanks @nik9000 .

Only nit is that I think the yaml tests should also run against the snapshot version, otherwise we never know if a change in snapshot invalidates an existing test.

This shouldn't hold this up, though, I'll try to address this in https://github.com/elastic/elasticsearch/pull/107433.

**Comment by alex-spies** [comment]
Update: since my PR already touches all YAML tests, I cherry-picked this into my [PR here](https://github.com/elastic/elasticsearch/pull/107433/commits/b394f4975d0885ab2fcc7756af6b4c6f9ecda781).

The difference is that my PR sets the version programmatically rather than pinning it in each individual YAML test.

**Comment by nik9000** [comment]
This was incorporated into another PR.

**Pull request closed without merging**